posts/2020-02-27-Rust-1.41.1.md

```diff
@@ -0,0 +1,128 @@
+a soundness hole related to static lifetimes, and a miscompilation causing segfaults.
+These regressions are not affecting past releases of Rust,
+and we recommend users of Rust 1.41.0 to upgrade as soon as possible.
+Another issue, related to interactions between `'static` and `Copy` implementations,
+dating back to Rust 1.0, was also addressed by this release.
+
+### A soundness hole in checking `static` items
+
+In Rust 1.41.0, due to some changes in the internal representation of `static` values,
+the borrow checker accidentally allowed some unsound programs.
+Specifically, the borrow checker would not check that `static` items had the correct type.
+This in turn would allow the assignment of a temporary,
+with a lifetime less than `'static`, to a `static` variable:
+
+```rust
+static mut MY_STATIC: &'static u8 = &0;
+
+fn main() {
+    let my_temporary = 42;
+    unsafe {
+        // Erroneously allowed in 1.41.0:
+        MY_STATIC = &my_temporary;
+    }
+}
+```
+
+This was addressed in 1.41.1, with the program failing to compile:
+```
+error[E0597]: `my_temporary` does not live long enough
+ --> src/main.rs:6:21
+  |
+6 |         MY_STATIC = &my_temporary;
+  |         ------------^^^^^^^^^^^^^
+  |         |           |
+  |         |           borrowed value does not live long enough
+  |         assignment requires that `my_temporary` is borrowed for `'static`
+7 |     }
+8 | }
+  | - `my_temporary` dropped here while still borrowed
+
+```
+
+You can learn more about this bug in [issue #69114][69114] and the [PR that fixed it][pr_69145].
+
+[69114]: https://github.com/rust-lang/rust/issues/69114
+[pr_69145]: https://github.com/rust-lang/rust/pull/69145
+
+### Respecting a `'static` lifetime in a `Copy` implementation
+
+[1.40.0_post]: https://blog.rust-lang.org/2019/12/19/Rust-1.40.0.html#borrow-check-migration-warnings-are-hard-errors-in-rust-2015
+
+Ever since Rust 1.0, the following erroneous program has been compiling:
+
+```rust
+#[derive(Clone)]
+struct Foo<'a>(&'a u32);
+impl Copy for Foo<'static> {}
+
+fn main() {
+    let temporary = 2;
+    let foo = (Foo(&temporary),);
+    drop(foo.0); // Accessing a part of `foo` is necessary.
+    drop(foo.0); // Indexing an array would also work.
+}
+```
+
+In Rust 1.41.1, this issue was fixed [by the same PR as the one above][pr_69145].
+Compiling the program now produces the following error:
+
+```rust
+
+error[E0597]: `temporary` does not live long enough
+  --> src/main.rs:7:20
+   |
+7  |     let foo = (Foo(&temporary),);
+   |                    ^^^^^^^^^^ borrowed value does not live long enough
+8  |     drop(foo.0);
+   |          ----- copying this value requires that
+   |                `temporary` is borrowed for `'static`
+9  |     drop(foo.0);
+10 | }
+   | - `temporary` dropped here while still borrowed
+```
+
+This happens because `Foo<'a>`, for some `'a`, only implements `Copy` when `'a: 'static`.
```

Comment:
```suggestion
This is an error because `Foo<'a>`, for some `'a`, only implements `Copy` when `'a: 'static`.
```

I thought "this happens because..." was going to explain the original cause of the bug; I think this is a clearer transition.

Reply:
```suggestion
This error occurs because `Foo<'a>`, for some `'a`, only implements `Copy` when `'a: 'static`.
```